### PR TITLE
added README.md for robot testtool

### PR DESCRIPTION
**Pull request by jomiklos**

### CODE REVIEW

File: confdgnmi/testtool/README.md [hunk 1]
```diff
@@ -0,0 +1,122 @@
+# gNMI Telemetry testing tool
+
+This project is intended as a tool to help with automation testing of a telemetry capabilities of target device.
+It uses a [Robot Framework](https://robotframework.org/) (a generic open source automation framework) to implement range of test suites and
+test cases. These try to identify capabilities of a device to service typical telemetry related gNMI requests.
+
+## requirements
+
+We strongly depend on **Python3** for running the tests. All the `python` commands in following text assume that this binary resolves to a (fairly recent) python3 interpreter. Following is an example of supported (not necessarily lowest required) version:
```
Comment (micnovak) [inline confdgnmi/testtool/README.md]: We require ...

File: confdgnmi/testtool/README.md [hunk 2]
```diff
@@ -0,0 +1,122 @@
+# gNMI Telemetry testing tool
+
+This project is intended as a tool to help with automation testing of a telemetry capabilities of target device.
+It uses a [Robot Framework](https://robotframework.org/) (a generic open source automation framework) to implement range of test suites and
+test cases. These try to identify capabilities of a device to service typical telemetry related gNMI requests.
+
+## requirements
```
Comment (martin-volf) [inline confdgnmi/testtool/README.md]: Do you really insist on lowercase section titles? It irritates my eyes a bit.

File: confdgnmi/testtool/README.md [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+It has a structure of a packed tree nodes that you can open to see specific test-suites and/or test-cases hierarchy, and read their names, descriptions, etc.
+
+If needed, see "`python -m robot.testdoc --help`" for possible output configuration.
+
+## running the tests
+
+To run any tests, we you need a target device configuration file first.
+
+When a test run is executed and finished, it creates few files by default (unless modified by `robot` command parameters).
+- `report.html` - good as a starting point, summary of all the tests suites/cases and their results;<br/>
```
Comment (micnovak) [inline confdgnmi/testtool/README.md]: I would remove " good as a starting point,"

File: confdgnmi/testtool/README.md [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    it links into the following `log.html` file for details...
+- `log.html` - detailed log for all the steps executed in test run
+
+### PYTHONPATH and executing tests
+
+Whether you run tests against "real" device, or the "ConfD adapter" (see further), you need properly set Python environment for tests to execute correctly.
```
Comment (micnovak) [inline confdgnmi/testtool/README.md:56]: Isn't "You need properly set Python environment for tests to execute correctly" enough?

File: confdgnmi/testtool/README.md [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+Report:  .../report.html
+```
+
+This results in creation of previously mentioned log files of interest - `report.html` and `log.html`.
+
+### running without target device - ConfD based "adapter"
+
+You can run tests without having any target device, e.g. to see an example test outputs/logs. You can do this by using the so-called "**ConfD gNMI adapter**" as a target device.
+
+ConfD gNMI adapter serves as an proof-of-concept implementation of gNMI enabled device. It utilizes the ConfD to run and service the model-specific requests for some of OpenConfig standard models<span style="background-color:red">, and TODO...?</span> It includes a simple gNMI server written in Python to "proxy' the incoming gNMI requests into the ConfD database.
```
Comment (micnovak) [inline confdgnmi/testtool/README.md]: And DEMO version for general gNMI tests.

File: confdgnmi/testtool/README.md [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+<span style="background-color:red">TODO - some seq. diagram could be useful here?</span
+
+Previously mentioned `adapter.yaml` file already contains all the variables settings needed for this use-case.
+
+<span style="background-color:red">TODO - add info on starting ConfD using Makefile (or add some helper script?) etc.</span>
```
Comment (micnovak) [inline confdgnmi/testtool/README.md]: Refer to adapter documentation (in docs)

File: confdgnmi/testtool/README.md [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+```
+
+See Robot's [installation instructions](https://robotframework.org/robotframework/latest/RobotFrameworkUserGuide.html#installation-instructions) for details on how to install the framework. To verify you have a `robot` installed, you can run `robot --version` shell command. As of writing this document, current version is:
+
+```bash
+$ robot --version
+Robot Framework 6.0.2 (Python 3.8.10 on linux)
+```
+
+Actual test-runs require that supplementary gNMI client code (located in `../src`, primarily `confd_gnmi_client.py` module) can be run successfully. See the `../README.md` for requirements if you run into some issues.
```
Comment (martin-volf) [inline confdgnmi/testtool/README.md:23]: The README file is just a pointer to the .adoc document, and that describes a lot more than just how to generate the stubs and set the Python path, which is everything you need to run the client.

File: confdgnmi/testtool/README.md [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+
+Assuming you have all the [requirements](#requirements) installed, just execute following shell command from within the current working directory (the one where this README.md file is located):
+
+```bash
+python -m robot.testdoc ./ TestDocumentation.html
+```
+
+This creates a single HTML file with specified name - that includes descriptions and details of all the test suites and test-cases from the project's `.robot` files.
+
+It has a structure of a packed tree nodes that you can open to see specific test-suites and/or test-cases hierarchy, and read their names, descriptions, etc.
```
Comment (martin-volf) [inline confdgnmi/testtool/README.md]: Is the adjective indeed "packed"? Perhaps "collapsed", or "folded"?

File: confdgnmi/testtool/README.md (hunk 1 above)
Comment (jomiklos) [reply confdgnmi/testtool/README.md]: fixed...

File: confdgnmi/testtool/README.md (hunk 3 above)
Comment (jomiklos) [reply confdgnmi/testtool/README.md]: removed...

File: confdgnmi/testtool/README.md [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+    it links into the following `log.html` file for details...
+- `log.html` - detailed log for all the steps executed in test run
+
+### PYTHONPATH and executing tests
+
+Whether you run tests against "real" device, or the "ConfD adapter" (see further), you need properly set Python environment for tests to execute correctly.
+
+<span style="background-color:red">TODO - think of some executor script to save user from having to declare PYTHONPATH?</span>
+
+For our caste, this implies setting the **`PYTHONPATH`** variable for underlying python code to find all the required items spread across various directories of the project. Assuming you run the tests from current directory, you need to have following set for tests to run:
```
Comment (martin-volf) [inline confdgnmi/testtool/README.md]: No castes please.

File: confdgnmi/testtool/README.md [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+
+### running against target device
+
+<span style="background-color:red">TODO -
+ maybe add a recommendation to verify target device connectivity from the shell before running tests to save some time/support effort?
+</span>
+
+To run the test cases against live target device, you need to write up your configuration file (e.g. "`MY.yaml`"). See `adapter.yaml` for an example of variables that need to be included in the file and their description. Create your own `.yaml` file that you can use in next step for executing the tests.
+
+as a parameter in command below to run the test suites:
```
Comment (martin-volf) [inline confdgnmi/testtool/README.md]: Runaway sentence?

File: confdgnmi/testtool/README.md [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+
+This results in creation of previously mentioned log files of interest - `report.html` and `log.html`.
+
+### running without target device - ConfD based "adapter"
+
+You can run tests without having any target device, e.g. to see an example test outputs/logs. You can do this by using the so-called "**ConfD gNMI adapter**" as a target device.
+
+ConfD gNMI adapter serves as an proof-of-concept implementation of gNMI enabled device. It utilizes the ConfD to run and service the model-specific requests for some of OpenConfig standard models<span style="background-color:red">, and TODO...?</span> It includes a simple gNMI server written in Python to "proxy' the incoming gNMI requests into the ConfD database.
+
+<span style="background-color:red">TODO - some seq. diagram could be useful here?</span
```
Comment (martin-volf) [inline confdgnmi/testtool/README.md]: I don't think so - here it makes sense to refer to the .adoc document.

File: confdgnmi/testtool/README.md [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+For our caste, this implies setting the **`PYTHONPATH`** variable for underlying python code to find all the required items spread across various directories of the project. Assuming you run the tests from current directory, you need to have following set for tests to run:
+
+```bash
+export PYTHONPATH=../src:./:./gNMI_Interface
+```
+
+### running against target device
+
+<span style="background-color:red">TODO -
+ maybe add a recommendation to verify target device connectivity from the shell before running tests to save some time/support effort?
```
Comment (martin-volf) [inline confdgnmi/testtool/README.md]: Perhaps just run `sanity` tests?

File: confdgnmi/testtool/README.md (hunk 2 above)
Comment (jomiklos) [reply confdgnmi/testtool/README.md]: was no explicit intention, will change to Capital... :)

File: confdgnmi/testtool/README.md (hunk 8 above)
Comment (jomiklos) [reply confdgnmi/testtool/README.md]: good point, will fix...

File: confdgnmi/testtool/README.md (hunk 9 above)
Comment (jomiklos) [reply confdgnmi/testtool/README.md]: good catch - typo, should have been "case" :)

File: confdgnmi/testtool/README.md (hunk 10 above)
Comment (jomiklos) [reply confdgnmi/testtool/README.md]: good catch, mistakenly deleted beginning of a sentence, will fix...

File: confdgnmi/testtool/README.md (hunk 12 above)
Comment (jomiklos) [reply confdgnmi/testtool/README.md]: good tip, will add...

File: confdgnmi/testtool/README.md (hunk 11 above)
Comment (jomiklos) [reply confdgnmi/testtool/README.md]: sounds good, will add...

File: confdgnmi/testtool/README.md (hunk 7 above)
Comment (jomiklos) [reply confdgnmi/testtool/README.md:23]: will add clarifying comment...

File: confdgnmi/testtool/README.md (hunk 6 above)
Comment (jomiklos) [reply confdgnmi/testtool/README.md]: will add...

File: confdgnmi/testtool/README.md (hunk 5 above)
Comment (jomiklos) [reply confdgnmi/testtool/README.md]: added info about demo mode...

File: confdgnmi/testtool/README.md [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+### PYTHONPATH and executing tests
+
+Whether you run tests against "real" device, or the "ConfD adapter" (see further), you need properly set Python environment for tests to execute correctly.
+
+<span style="background-color:red">TODO - think of some executor script to save user from having to declare PYTHONPATH?</span>
+
+For our caste, this implies setting the **`PYTHONPATH`** variable for underlying python code to find all the required items spread across various directories of the project. Assuming you run the tests from current directory, you need to have following set for tests to run:
+
+```bash
+export PYTHONPATH=../src:./:./gNMI_Interface
```
Comment (micnovak) [inline confdgnmi/testtool/README.md]: Another option can be 

`PYTHONPATH=../src:./:./gNMI_Interface robot --variablefile  ....`

File: confdgnmi/testtool/README.md [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+### running without target device - ConfD based "adapter"
+
+You can run tests without having any target device, e.g. to see an example test outputs/logs. You can do this by using the so-called "**ConfD gNMI adapter**" as a target device.
+
+ConfD gNMI adapter serves as an proof-of-concept implementation of gNMI enabled device. It utilizes the ConfD to run and service the model-specific requests for some of OpenConfig standard models<span style="background-color:red">, and TODO...?</span> It includes a simple gNMI server written in Python to "proxy' the incoming gNMI requests into the ConfD database.
+
+<span style="background-color:red">TODO - some seq. diagram could be useful here?</span
+
+Previously mentioned `adapter.yaml` file already contains all the variables settings needed for this use-case.
+
```
Comment (micnovak) [inline confdgnmi/testtool/README.md:123]: Add few copy paste commands, so reader can try right away
e.g.

```
make clean all start`
./src/confd_gnmi_server.py -t api  --insecure 
```
or 

```
make clean gnmi_proto 
./src/confd_gnmi_server.py -t demo  --insecure 
```

and in other terminal 

`PYTHONPATH=../src:./:./gNMI_Interface robot --variablefile adapter.yaml --include sanity ./`

Ideally for both DEMO (for those not having access to ConfD) and API adapter

File: confdgnmi/testtool/README.md (hunk 13 above)
Comment (jomiklos) [reply confdgnmi/testtool/README.md]: correct, i will update this part to hint on both options

File: confdgnmi/testtool/README.md (hunk 14 above)
Comment (jomiklos) [reply confdgnmi/testtool/README.md:123]: good point, added...

**Pull request merged**